android/app/src/main/cpp/app/organicmaps/bookmarks/data/BookmarkManager.cpp

```diff
@@ -188,16 +188,28 @@ void OnPreparedFileForSharing(JNIEnv * env, BookmarkManager::SharingResult const
 
   static jclass const classBookmarkSharingResult = jni::GetGlobalClassRef(env,
     "app/organicmaps/bookmarks/data/BookmarkSharingResult");
-  // Java signature : BookmarkSharingResult(long categoryId, @Code int code,
+  // Java signature : BookmarkSharingResult(List<Long> categoriesIds, @Code int code,
```

Comment:
I thought about it initially, when I saw that there is already implementation for array. Later I had an idea that list/vector could be more convenient on java/cpp side while conversion code is implemented only once. If there are performance concerns (I don't know how much difference is if we pass only collections (not specific bookmarks/tracks) here).

Reply:
```
  jsize const size = static_cast<jsize>(result.m_categoriesIds.size());
  jni::ScopedLocalRef<jlongArray> jCategoriesIds(env, env->NewLongArray(size));
  env->SetLongArrayRegion(jResults.get(), 0, size, result.m_categoriesIds.data());
```
should be easier and faster.

Reply:
> ```
>   jsize const size = static_cast<jsize>(result.m_categoriesIds.size());
>   jni::ScopedLocalRef<jlongArray> jCategoriesIds(env, env->NewLongArray(size));
>   env->SetLongArrayRegion(jResults.get(), 0, size, result.m_categoriesIds.data());
> ```
> 
> should be easier and faster.

hm, I don't understand the suggested code - in this function we receive cpp object, build java object BookmarkSharingResult and pass it to onPreparedFileForSharing. 
What is the jResult variable in this code snippet and how array jCategoriesIds is used?

Reply:
> ```
>   jsize const size = static_cast<jsize>(result.m_categoriesIds.size());
>   jni::ScopedLocalRef<jlongArray> jCategoriesIds(env, env->NewLongArray(size));
>   env->SetLongArrayRegion(jResults.get(), 0, size, result.m_categoriesIds.data());
> ```
> 
> should be easier and faster.

hm, I don't understand the suggested code - in this function we receive cpp object, build java object BookmarkSharingResult and pass it to onPreparedFileForSharing.
What is the jResult variable in this code snippet and how array jCategoriesIds is used?

Reply:
@cyber-toad this is an example of how to use `long[]` array instead of `List<Long>`. 
jResults is a typo, of course. It should be jCategoriesIds

Reply:
@biodranik I switched to long[] - there was an issue with SetLongArrayRegion/GetLongArrayRegion - compilation error because jlong is signed and MarkId is unsigned - I solved it with tmp vector as in android/app/src/main/cpp/app/organicmaps/SearchEngine.cpp:212. Please have a look.